drivers: add driver specific Makefile.dep

aabadie opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description

This PR splits the main `drivers/Makefile.dep` in driver specific `Makefile.dep`s.

The dependency resolution for drivers modules defined as pseudo-modules still have to be done in the main `drivers/Makefile.dep` but just add the "real" module as a dependency.

There was a dependency resolution for `nrfmin` in `drivers/Makefile.dep` but it has nothing to do here, this PR moves it to `cpu/nrf5x_common/Makefile.dep`.

<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->


### Testing procedure

- A green Murdock
- Use build system tools to ensure no dependency has changed (help: what are the commands already ?)

<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->


### Issues/PRs references

Similar to #14369 but for Makefile.dep

<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

aabadie commented:
I fear that this one will be hard to maintain in the long run...

gschorcht commented:
> * Use build system tools to ensure no dependency has changed (help: what are the commands already ?)

Would be `dist/tools/buildsystem_sanity_check/save_all_dependencies_resolution_variables.sh` sufficient?

fjmolinas commented:
> I fear that this one will be hard to maintain in the long run...

You mean this PR, or the actual changes? If no dependencies have changed I think we can move quickly with this one, but is @miri64 OK with us merging this one in soft feature freeze?

fjmolinas commented:
> You mean this PR, or the actual changes? If no dependencies have changed I think we can move quickly with this one, but is @miri64 OK with us merging this one in soft feature freeze?

I'm currently running the script for #14353 I can run i for this PR after wards.

gschorcht commented on drivers/Makefile.dep:
```diff
@@ -1,875 +1,129 @@
-  FEATURES_REQUIRED += periph_spi
 endif
 
 ifneq (,$(filter ccs811_full,$(USEMODULE)))
```
I wonder if it would be better to always address driver variants defined by pseudomodules with `%`, for example.
```suggestion
ifneq (,$(filter ccs811%,$(USEMODULE)))
```
For most of the pseudomodules it's already done in that way, but for some of them direct names are used.

aabadie replied:
Sure, I started with the brute force way but we can refine for sure !

gschorcht replied:
Since we touch it once, we should change it with this PR. If I'm right, there are less than 10 occurrences.

aabadie replied:
Addressed, I prefer to use `ccs811_%`.

gschorcht commented on drivers/mhz19/Makefile.dep:
```diff
@@ -0,0 +1,9 @@
+  USEMODULE += xtimer
```
```suggestion
USEMODULE += xtimer
```

aabadie replied:
Fixed

aabadie commented:
> I'm currently running the script for #14353 I can run i for this PR after wards.

Thanks!

gschorcht commented:
@aabadie I went through all the changes. From what I have seen, it should be the same. But we should rely on tools. If it helps, I could run the script if @fjmolinas run for PR #14353 takes too long.

fjmolinas commented:
> @aabadie I went through all the changes. From what I have seen, it should be the same. But we should rely on tools. If it helps, I could run the script if @fjmolinas run for PR #14353 takes too long.

The script is still running on my side, so feel free to run it yourself. I would probably run it tonight and look at the results tomorrow/

gschorcht commented:
> > @aabadie I went through all the changes. From what I have seen, it should be the same. But we should rely on tools. If it helps, I could run the script if @fjmolinas run for PR #14353 takes too long.
> 
> The script is still running on my side, so feel free to run it yourself. I would probably run it tonight and look at the results tomorrow/

Ok, I will start it immediatly. My computer is borrowing :wink:

gschorcht commented:
@fjmolinas for each application I get the two warnings
```
boards/arduino-nano-33-ble/Makefile.include:27: warning: overriding recipe for target 'term-delay'
makefiles/tools/usb_board_reset.mk:35: warning: ignoring old recipe for target 'term-delay'
```
Is that normal?

fjmolinas commented:
> @fjmolinas for each application I get the two warnings
> ```
> boards/arduino-nano-33-ble/Makefile.include:27: warning: overriding recipe for target 'term-delay'
> makefiles/tools/usb_board_reset.mk:35: warning: ignoring old recipe for target 'term-delay'
> ```
> Is that normal?

Not sure if normal, but unrelated. I'll take a look.

aabadie commented:
I switched back to the original way of performing the dependency resolution of nrfmin. The list of dependencies should be different.

gschorcht commented:
> I switched back to the original way of performing the dependency resolution of nrfmin. The list of dependencies should be different compared to the initial version of this PR.

I'm running the script on the former version of this PR. Let's see what happens.

aabadie commented:
I'm sorry, I had to push another fixup because I broke the bmx280 dependency resolution. Let's wait for Murdock to pass (except the statis test) before running the dependency check script.

fjmolinas commented:
The output of running `./dist/tools/buildsystem_sanity_check/save_all_dependencies_resolution_variables.sh` `[diff.txt](https://github.com/RIOT-OS/RIOT/files/4849885/diff.txt)`, there are only differences in `FEATURES_REQUIRED` and `FEATURES_OPTIONAL`, but no diff in `FEATURES_USED` or `USEMODULE`. So there is no dependencies have changed, it is probably parsed more times than before. I wan't to try to understand why, but I dont' necessarily see this as a problem.

gschorcht commented:
> The output of running `./dist/tools/buildsystem_sanity_check/save_all_dependencies_resolution_variables.sh` [diff.txt](https://github.com/RIOT-OS/RIOT/files/4849885/diff.txt), there are only differences in `FEATURES_REQUIRED` and `FEATURES_OPTIONAL`, but no diff in `FEATURES_USED` or `USEMODULE`. So there is no dependencies have changed, it is probably parsed more times than before. I wan't to try to understand why, but I dont' necessarily see this as a problem.

Wow, you have already finished it. My one is still running since I forgot to start the second run yesterday in the evening.

fjmolinas commented:
> > The output of running `./dist/tools/buildsystem_sanity_check/save_all_dependencies_resolution_variables.sh` [diff.txt](https://github.com/RIOT-OS/RIOT/files/4849885/diff.txt), there are only differences in `FEATURES_REQUIRED` and `FEATURES_OPTIONAL`, but no diff in `FEATURES_USED` or `USEMODULE`. So there is no dependencies have changed, it is probably parsed more times than before. I wan't to try to understand why, but I dont' necessarily see this as a problem.
> 
> Wow, you have already finished it. My one is still running since I forgot to start the second run yesterday in the evening :worried: 

I have a second machine, useful for running this kind of things while keeping my laptop free.

gschorcht commented:
> The output of running `./dist/tools/buildsystem_sanity_check/save_all_dependencies_resolution_variables.sh` [diff.txt](https://github.com/RIOT-OS/RIOT/files/4849885/diff.txt), there are only differences in `FEATURES_REQUIRED` and `FEATURES_OPTIONAL`

`FEATURES_REQUIRED` and `FEATURES_OPTIONAL` look really weird at times. But as I understand it, it is important that the sorted versions of these lists do not differ.

gschorcht commented:
Now that the release branch was created after the hard feature freeze, we should merge this PR as soon as possible before new drivers are merged.

@aabadie Please squash.
@fjmolinas Based on the results of your comparison in https://github.com/RIOT-OS/RIOT/pull/14386#issuecomment-651582152, can we be sure that the dependencies are resolved correctly, even if `FEATURES_REQUIRED` and `FEATURES_OPTIONAL` sometimes look very strange (the complete list of features multiple times)? Sorted versions of these variables don't differ.

aabadie commented:
I had conflicts when rebasing the last time (a couple of hours ago) so I think it's safer to re-run the checks. I rebased again and squashed the branch.

aabadie commented:
rebased once again.

gschorcht commented:
Again, looks good from my point of view.

Ping @fjmolinas Based on the results of your comparison in https://github.com/RIOT-OS/RIOT/pull/14386#issuecomment-651582152, can we be sure that the dependencies are resolved correctly, even if `FEATURES_REQUIRED` and `FEATURES_OPTIONAL` sometimes look very strange (the complete list of features is in multiple times)? Sorted versions of these variables don't differ.

aabadie commented:
rebased to solve the conflict introduced when #14497 was just merged.

gschorcht commented:
We should merge it before further conflicts pop up. But I'm a bit unsure without the final statement from @fjmolinas :worried:

aabadie commented:
> But I'm a bit unsure without the final statement from @fjmolinas

I can ping him IRL :D

gschorcht commented:
> > But I'm a bit unsure without the final statement from @fjmolinas
> 
> I can ping him IRL :D

That would be good. My concern was that although the sorted versions of `FEATURES_REQUIRED` and `FEATURES_OPTIONAL` were the same in his test, the unsorted versions of these variables sometimes looked really strange.

fjmolinas commented on Makefile.dep:
```diff
@@ -15,6 +15,9 @@
 include $(RIOTBASE)/drivers/Makefile.dep
 
+# pull Makefile.dep of each driver modules if they exist
+-include $(USEMODULE:%=$(RIOTBASE)/drivers/%/Makefile.dep)
```
```suggestion
-include $(sort $(USEMODULE:%=$(RIOTBASE)/drivers/%/Makefile.dep))
```

This will fix having the list included multiple times.

aabadie replied:
Addressed

Nishchay-sopho commented:
The driver-guide.md file is not updated to include these changes since it still says that you have to add the driver dependencies in the Makefile.dep file in `$(RIOTBASE)/drivers/Makefile.dep`
